### PR TITLE
Commutative undefined minify

### PR DESCRIPTION
**Pull request by bmeck**
this should probably be squashed, i'm not familiar with code commit norms here. we can expand a bunch of the commutative compressions later.

### CODE REVIEW

File: crates/oxc_minifier/src/compressor/mod.rs [hunk 1]
```diff
@@ -203,20 +203,77 @@ impl<'a> Compressor<'a> {
...
+        //                     let left = self.ast.identifier_reference_expression((*ident).clone());
+        //                     let right = self.ast.void_0();
+        //                     let operator = BinaryOperator::StrictEquality;
+        //                     *expr = BinaryExpression { span: SPAN, left, operator, right };
+        //                 }
+        //             }
+        //         }
+        //     }
+        // }
+        // return
```
Comment (Boshen) [inline crates/oxc_minifier/src/compressor/mod.rs:220]: Let's remove the commented out code 😅

File: crates/oxc_minifier/src/compressor/mod.rs [hunk 2]
```diff
@@ -203,20 +203,77 @@ impl<'a> Compressor<'a> {
...
+                    };
+                    *expr = cmp;
                 }
+
+            }
+            _ => {}
+        };
+    }
+
+    pub fn commutative_pair<A, F, G, RetF: 'a, RetG: 'a>(
```
Comment (Boshen) [inline crates/oxc_minifier/src/compressor/mod.rs:261]:
```suggestion
   fn commutative_pair<A, F, G, RetF: 'a, RetG: 'a>(
```

**Review by Boshen — CHANGES_REQUESTED** [review]

File: crates/oxc_minifier/src/compressor/mod.rs (hunk 1 above)
Comment (Boshen) [inline crates/oxc_minifier/src/compressor/mod.rs:220]:
```suggestion
```

**Review by Boshen — APPROVED** [review]

File: crates/oxc_minifier/src/compressor/mod.rs [hunk 3]
```diff
@@ -203,20 +203,62 @@ impl<'a> Compressor<'a> {
...
-                    if let Expression::Identifier(ident) = &unary_expr.argument {
-                        if expr.right.is_specific_string_literal("undefined") {
-                            let left = self.ast.identifier_reference_expression((*ident).clone());
-                            let right = self.ast.void_0();
-                            let operator = BinaryOperator::StrictEquality;
-                            *expr = BinaryExpression { span: SPAN, left, operator, right };
+        if !self.options.typeofs {
+            return
+        }
+        // return None;
```
Comment (Boshen) [inline crates/oxc_minifier/src/compressor/mod.rs:209]:
```suggestion
```

**Comment by Boshen** [comment]
Thank you! A great start on learning Rust!

**Review by Boshen — COMMENTED** [review]
Need a quick `cargo clippy` and `cargo format` fix.

File: crates/oxc_minifier/tests/oxc/folding.rs [hunk 4]
```diff
@@ -8,6 +8,12 @@ fn addition_folding() {
     test("x+''", "x+'';");
 }
 
+#[test]
+fn typeof_folding() {
+    test("typeof x === 'undefined'", "void 0===x;");
```
Comment (magic-akari) [inline crates/oxc_minifier/tests/oxc/folding.rs:13]: This is dangerous. People use `typeof x` to determine if `x` exists.
For examples:

```JavaScript
if(typeof process === 'undefined') {
// not in node.js
}
```

If the definition of `x` is found in the context, then such an operation is feasible.

```JavaScript
function foo(x) {
    typeof x === 'undefined'; // We know that the variable x definitely exists.
}
```

**Comment by bmeck** [comment]
@Boshen I've run those cmds and uploaded the new commit

**Pull request merged**